src/sas/sascalc/fit/BumpsFitting.py

```diff
@@ -282,77 +282,112 @@ def fit(self, msg_q=None,
+            par_names = fitness.fitted_par_names + fitness.computed_par_names
+
             R = FResult(model=fitness.model,
                         data=fitness.data,
                         param_list=par_names)
             R.theory = fitness.theory()
             R.residuals = fitness.residuals()
             R.index = fitness.data.idx
             R.fitter_id = self.fitter_id
+            # TODO: should scale stderr by sqrt(chisq/DOF) if dy is unknown
             R.success = result['success']
+            R.convergence = result['convergence']
+            if result['uncertainty'] is not None:
+                R.uncertainty_state = result['uncertainty']
+
             if R.success:
-                R.pvec = (np.array([p.value.n for p in pars]) if
-                    propagate_errors else result['value'])
-                R.stderr = (np.array([p.value.s for p in pars]) if
-                    propagate_errors else result["stderr"])
+                pvec = list()
+                stderr = list()
+                for p in pars:
+                    # If p is already defined as an uncertainties object it is not constrained based on another
+                    # parameter
+                    if isinstance(p.value, uncertainties.core.Variable) or \
+                            isinstance(p.value, uncertainties.core.AffineScalarFunc):
+                        # value.n returns value p
+                        pvec.append(p.value.n)
+                        # value.n returns error in p
+                        stderr.append(p.value.s)
+                    # p constrained based on another parameter
+                    else:
+                        # Details of p
+                        param_model, param_name = p.name.split(".")[0], p.name.split(".")[1]
+                        # Constraints applied on p, list comprehension most efficient method, will always return a
+                        # list with 1 entry
+                        constraints = [model.constraints for model in models if model.name == param_model][0]
+                        # Parameters p is constrained on.
+                        reference_params = [v for v in varying if str(v.name) in str(constraints[param_name])]
+                        err_exp = str(constraints[param_name])
+                        # Convert string entries into variable names within the code.
+                        for i in enumerate(reference_params):
+                            index = i[0]
+                            err_exp = err_exp.replace(reference_params[index].name, f"reference_params[{index}].value")
+                        try:
+                            # Evaluate a string containing constraints as if it where a line of code
+                            pvec.append(eval(err_exp).n)
+                            stderr.append(eval(err_exp).s)
+                        except NameError as e:
+                            # Get model causing error
```

Comment:
I think `eval` returns all sorts of exceptions, also e.g. `SyntaxError` - are you sure these can't be triggered by `err_exp`?
Maybe we need a blanket `except Exception as e:` instead?

Reply:
That's a good point, I'm pretty sure Syntax error will not occur, as you are limited when writing a constraint manually. But I will add a second exception catcher in the case of an unexpected error that will affect the error propagation, but not the plotting

This Exception is caused when a user has a constraint on for a parameter that hasn't been calculated. For example:
M1.radius = M2.radius * 2
M3.radius = M2.radius + M1.radius 
is fine but
M3.radius = M2.radius + M1.radius 
M1.radius = M2.radius * 2
will lead to an error, as m3 is constrained based on the value for m1, but this hasn't been calculated yet as there is a constraint on m1.

This error could be fixed by looping over all the constraints multiple times, until the prerequisites for each parameter has been computed. However this hasn't been added as it's easy for the user to avoid the issue, by reordering the constraints.

Reply:
Are we allowing iterative, stacked up, constraints at all in the constraints set up? I thought not ????
( I have made much use of them myself in FISH, but you do have to be careful as to the order in which they are applied, e.g I suspect sorting the error bars out might require running through the list of constraints in reverse order?)